File: RxTools-library/src/main/java/com/vondear/rxtools/view/RxRotateBarBasic.java
package com.vondear.rxtools.view;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.text.TextPaint;

import java.util.ArrayList;

/**
 *
 * @author Vondear
 * @date 12/24/15
 */
public class RxRotateBarBasic {

    private static final int ITEM_OFFSET = 1;

    private static final int SHADOW_ALPHA = (int) (255 * 0.2); // 20%

    private static final int UN_RATING_ALPHA = (int) (255 * 0.3); // 30%

    private static final int RATING_ALPHA = (int) (255 * 1.0); // 100%

    private static final int OUTLINE_ALPHA = (int) (255 * 0.4); // 40%

    private boolean isSingle = false;

    private boolean isShowTitle = true;

    private int outlineWidth, ratingBarWidth, shadowWidth, textWidth;

    private float mStartAngle;

    private float mSweepAngle;

    private int maxRate = 10;

    private int mCurRate;

    private ArrayList<Rate> rates;

    private Paint ratedPaint, unRatedPaint, shadowPaint, outlinePaint;
    private TextPaint titlePaint;

    private int mRadius;

    private String mTitle;

    private int mCenterX, mCenterY;

    private RectF outlineOval, ratingOval, shadowOval;

    public RxRotateBarBasic(int curRate, String title) {
        this.mCurRate = curRate;
        this.mTitle = title;

        outlinePaint = new Paint();
        ratedPaint = new Paint();
        unRatedPaint = new Paint();
        shadowPaint = new Paint();
        titlePaint = new TextPaint();
        // set default color
        setRatingBarColor(Color.WHITE);
    }

    protected void init() {
        // in this order to init
        initRatingBar();
        initOval();
        initPaint();
    }

    private void initOval() {
        mRadius = mCenterX > mCenterY ? mCenterY : mCenterX;

        // text bar : 1/10 of radius
        textWidth = mRadius / 10;
        // rating bar : 1/10 of radius
        ratingBarWidth = mRadius / 10;
        // shadow : 1/3 of rating bar
        shadowWidth = ratingBarWidth / 3;
        // outline : 1/3 of shadow
        outlineWidth = shadowWidth / 3;

        int outlineRadius = mRadius - (textWidth / 2); // outline include text and outline, radius is base on textWidth

        int paddingRadius = outlineRadius - (textWidth / 2); // padding space draw nothing, radius is base on textWidth

        int ratingBarRadius = paddingRadius - (textWidth / 2) - (ratingBarWidth / 2);

        int shadowRadius = ratingBarRadius - (ratingBarWidth / 2) - (shadowWidth / 2);

        outlineOval = new RectF();
        ratingOval = new RectF();
        shadowOval = new RectF();

        outlineOval.left = mCenterX - outlineRadius;
        outlineOval.top = mCenterY - outlineRadius;
        outlineOval.right = mCenterX + outlineRadius;
        outlineOval.bottom = mCenterY + outlineRadius;

        ratingOval.left = mCenterX - ratingBarRadius;
        ratingOval.top = mCenterY - ratingBarRadius;
        ratingOval.right = mCenterX + ratingBarRadius;
        ratingOval.bottom = mCenterY + ratingBarRadius;

        shadowOval.left = mCenterX - shadowRadius;
        shadowOval.top = mCenterY - shadowRadius;
        shadowOval.right = mCenterX + shadowRadius;
        shadowOval.bottom = mCenterY + shadowRadius;
    }

    private void initPaint() {

        outlinePaint.setAntiAlias(true);
        outlinePaint.setStyle(Paint.Style.STROKE);
        outlinePaint.setStrokeWidth(outlineWidth);
        outlinePaint.setAlpha(OUTLINE_ALPHA);

        ratedPaint.setAntiAlias(true);
        ratedPaint.setStyle(Paint.Style.STROKE);
        ratedPaint.setStrokeWidth(ratingBarWidth);
        ratedPaint.setAlpha(RATING_ALPHA);

        unRatedPaint.setAntiAlias(true);
        unRatedPaint.setStyle(Paint.Style.STROKE);
        unRatedPaint.setStrokeWidth(ratingBarWidth);
        unRatedPaint.setAlpha(UN_RATING_ALPHA);

        shadowPaint.setAntiAlias(true);
        shadowPaint.setStyle(Paint.Style.STROKE);
        shadowPaint.setStrokeWidth(shadowWidth);
        shadowPaint.setAlpha(SHADOW_ALPHA);

        titlePaint.setAntiAlias(true);
        titlePaint.setTextSize(textWidth);
        titlePaint.setAlpha(RATING_ALPHA);
    }

    private void initRatingBar() {
        rates = new ArrayList<>();

        float itemSweepAngle;
        if (isSingle) {
            itemSweepAngle = (mSweepAngle - (ITEM_OFFSET * (maxRate))) / maxRate;
        } else {
            itemSweepAngle = (mSweepAngle - (ITEM_OFFSET * (maxRate - 1))) / maxRate;
        }

        for (int i = 0; i < maxRate; i++) {
            float itemStartAngle = mStartAngle + i * (itemSweepAngle + ITEM_OFFSET);
            rates.add(new Rate(itemStartAngle, itemSweepAngle));
        }
    }

    protected void drawUnRate(Canvas canvas) {
        for (Rate arc : rates) {
            arc.drawArc(canvas, ratingOval, unRatedPaint);
        }
    }

    protected void drawRate(Canvas canvas, int index) {
        if (index >= maxRate) {
            return;
        }
        Rate arc = rates.get(index);
        arc.drawArc(canvas, ratingOval, ratedPaint);
    }

    protected void drawShadow(Canvas canvas) {
        for (Rate arc : rates) {
            arc.drawArc(canvas, shadowOval, shadowPaint);
        }
    }

    protected void drawTitle(Canvas canvas, int alpha) {
        if (alpha > 0 && isShowTitle) {
            Path path = new Path();
            float circumference = (float) (Math.PI * (outlineOval.right - outlineOval.left));
            float textAngle = (360 / circumference) * titlePaint.measureText(getTitle());

            float startAngle = mStartAngle + mSweepAngle / 2 - textAngle / 2;

            if (isSingle) {
                // when single, draw 360 the path will be a circle
                path.addArc(outlineOval, startAngle - mSweepAngle / 2, mSweepAngle / 2);
            } else {
                path.addArc(outlineOval, startAngle, mSweepAngle);
            }

            titlePaint.setAlpha(alpha);
            canvas.drawTextOnPath(mTitle, path, 0, textWidth / 3, titlePaint);
        }
    }

    protected void drawOutLine(Canvas canvas) {

        float circumference = (float) (Math.PI * (outlineOval.right - outlineOval.left));
        float textAngle = (360 / circumference) * titlePaint.measureText(getTitle());

        float sweepAngle = (mSweepAngle - textAngle - 1 - 1) / 2;

        if (isShowTitle) {
            // text left
            float leftStartAngle = mStartAngle;
            canvas.drawArc(outlineOval, leftStartAngle, sweepAngle, false, outlinePaint);
            // text right
            float rightStartAngle = mStartAngle + mSweepAngle - sweepAngle;
            canvas.drawArc(outlineOval, rightStartAngle, sweepAngle, false, outlinePaint);
        } else {
            canvas.drawArc(outlineOval, mStartAngle, mSweepAngle, false, outlinePaint);
        }


    }

    public void setMaxRate(int maxRate) {
        this.maxRate = maxRate;
    }

    public int getRate() {
        return mCurRate;
    }

    public void setRate(int curRate) {
        this.mCurRate = curRate;
    }

    protected void setStartAngle(float mStartAngle) {
        this.mStartAngle = mStartAngle;
    }

    protected void setSweepAngle(float mSweepAngle) {
        this.mSweepAngle = mSweepAngle;
    }

    protected void setCenterX(int mCenterX) {
        this.mCenterX = mCenterX;
    }

    protected void setCenterY(int mCenterY) {
        this.mCenterY = mCenterY;
    }

    protected void setIsSingle(boolean isSingle) {
        this.isSingle = isSingle;
    }

    public String getTitle() {
        return mTitle;
    }

    public void setTitle(String title) {
        this.mTitle = title;
    }

    public void setRatedColor(int color) {
        ratedPaint.setColor(color);
    }

    public void setUnRatedColor(int color) {
        unRatedPaint.setColor(color);
    }

    public void setTitleColor(int color) {
        titlePaint.setColor(color);
    }

    public void setOutlineColor(int color) {
        outlinePaint.setColor(color);
    }

    public void setShadowColor(int color) {
        shadowPaint.setColor(color);
    }

    public void isShowTitle(boolean isShow) {
        isShowTitle = isShow;
    }

    public void setRatingBarColor(int color) {
        ratedPaint.setColor(color);
        unRatedPaint.setColor(color);
        titlePaint.setColor(color);
        outlinePaint.setColor(color);
        shadowPaint.setColor(color);
    }

    /**
     * Rate class
     */
    public class Rate {
        private float startAngle, sweepAngle;

        public Rate(float startAngle, float sweepAngle) {
            this.startAngle = startAngle;
            this.sweepAngle = sweepAngle;
        }

        public void drawArc(Canvas canvas, RectF oval, Paint paint) {
            canvas.drawArc(oval, startAngle, sweepAngle, false, paint);
        }
    }

}

File: RxUI/src/main/java/com/tamsiree/rxui/view/RxRotateBarBasic.kt
package com.tamsiree.rxui.view

import android.graphics.*
import android.text.TextPaint
import java.util.*

/**
 * @author Tamsiree
 * @date 12/24/15
 */
class RxRotateBarBasic(var rate: Int, var title: String) {
    private var isSingle = false
    private var isShowTitle = true
    private var outlineWidth = 0
    private var ratingBarWidth = 0
    private var shadowWidth = 0
    private var textWidth = 0
    private var mStartAngle = 0f
    private var mSweepAngle = 0f
    private var maxRate = 10
    private var rates: ArrayList<Rate>? = null
    private val ratedPaint: Paint = Paint()
    private val unRatedPaint: Paint = Paint()
    private val shadowPaint: Paint = Paint()
    private val outlinePaint: Paint = Paint()
    private val titlePaint: TextPaint = TextPaint()
    private var mRadius = 0
    private var mCenterX = 0
    private var mCenterY = 0
    private var outlineOval: RectF? = null
    private var ratingOval: RectF? = null
    private var shadowOval: RectF? = null
    fun init() {
        // in this order to init
        initRatingBar()
        initOval()
        initPaint()
    }

    private fun initOval() {
        mRadius = if (mCenterX > mCenterY) mCenterY else mCenterX

        // text bar : 1/10 of radius
        textWidth = mRadius / 10
        // rating bar : 1/10 of radius
        ratingBarWidth = mRadius / 10
        // shadow : 1/3 of rating bar
        shadowWidth = ratingBarWidth / 3
        // outline : 1/3 of shadow
        outlineWidth = shadowWidth / 3

        // outline include text and outline, radius is base on textWidth
        val outlineRadius = mRadius - textWidth / 2

        // padding space draw nothing, radius is base on textWidth
        val paddingRadius = outlineRadius - textWidth / 2
        val ratingBarRadius = paddingRadius - textWidth / 2 - ratingBarWidth / 2
        val shadowRadius = ratingBarRadius - ratingBarWidth / 2 - shadowWidth / 2
        outlineOval = RectF()
        ratingOval = RectF()
        shadowOval = RectF()
        outlineOval!!.left = mCenterX - outlineRadius.toFloat()
        outlineOval!!.top = mCenterY - outlineRadius.toFloat()
        outlineOval!!.right = mCenterX + outlineRadius.toFloat()
        outlineOval!!.bottom = mCenterY + outlineRadius.toFloat()
        ratingOval!!.left = mCenterX - ratingBarRadius.toFloat()
        ratingOval!!.top = mCenterY - ratingBarRadius.toFloat()
        ratingOval!!.right = mCenterX + ratingBarRadius.toFloat()
        ratingOval!!.bottom = mCenterY + ratingBarRadius.toFloat()
        shadowOval!!.left = mCenterX - shadowRadius.toFloat()
        shadowOval!!.top = mCenterY - shadowRadius.toFloat()
        shadowOval!!.right = mCenterX + shadowRadius.toFloat()
        shadowOval!!.bottom = mCenterY + shadowRadius.toFloat()
    }

    private fun initPaint() {
        outlinePaint.isAntiAlias = true
        outlinePaint.style = Paint.Style.STROKE
        outlinePaint.strokeWidth = outlineWidth.toFloat()
        outlinePaint.alpha = OUTLINE_ALPHA
        ratedPaint.isAntiAlias = true
        ratedPaint.style = Paint.Style.STROKE
        ratedPaint.strokeWidth = ratingBarWidth.toFloat()
        ratedPaint.alpha = RATING_ALPHA
        unRatedPaint.isAntiAlias = true
        unRatedPaint.style = Paint.Style.STROKE
        unRatedPaint.strokeWidth = ratingBarWidth.toFloat()
        unRatedPaint.alpha = UN_RATING_ALPHA
        shadowPaint.isAntiAlias = true
        shadowPaint.style = Paint.Style.STROKE
        shadowPaint.strokeWidth = shadowWidth.toFloat()
        shadowPaint.alpha = SHADOW_ALPHA
        titlePaint.isAntiAlias = true
        titlePaint.textSize = textWidth.toFloat()
        titlePaint.alpha = RATING_ALPHA
    }

    private fun initRatingBar() {
        rates = ArrayList()
        val itemSweepAngle: Float
        itemSweepAngle = if (isSingle) {
            (mSweepAngle - ITEM_OFFSET * maxRate) / maxRate
        } else {
            (mSweepAngle - ITEM_OFFSET * (maxRate - 1)) / maxRate
        }
        for (i in 0 until maxRate) {
            val itemStartAngle = mStartAngle + i * (itemSweepAngle + ITEM_OFFSET)
            rates!!.add(Rate(itemStartAngle, itemSweepAngle))
        }
    }

    fun drawUnRate(canvas: Canvas) {
        for (arc in rates!!) {
            arc.drawArc(canvas, ratingOval, unRatedPaint)
        }
    }

    fun drawRate(canvas: Canvas, index: Int) {
        if (index >= maxRate) {
            return
        }
        val arc = rates!![index]
        arc.drawArc(canvas, ratingOval, ratedPaint)
    }

    fun drawShadow(canvas: Canvas) {
        for (arc in rates!!) {
            arc.drawArc(canvas, shadowOval, shadowPaint)
        }
    }

    fun drawTitle(canvas: Canvas, alpha: Int) {
        if (alpha > 0 && isShowTitle) {
            val path = Path()
            val circumference = (Math.PI * (outlineOval!!.right - outlineOval!!.left)).toFloat()
            val textAngle = 360 / circumference * titlePaint.measureText(title)
            val startAngle = mStartAngle + mSweepAngle / 2 - textAngle / 2
            if (isSingle) {
                // when single, draw 360 the path will be a circle
                path.addArc(outlineOval, startAngle - mSweepAngle / 2, mSweepAngle / 2)
            } else {
                path.addArc(outlineOval, startAngle, mSweepAngle)
            }
            titlePaint.alpha = alpha
            canvas.drawTextOnPath(title, path, 0f, textWidth / 3.toFloat(), titlePaint)
        }
    }

    fun drawOutLine(canvas: Canvas) {
        val circumference = (Math.PI * (outlineOval!!.right - outlineOval!!.left)).toFloat()
        val textAngle = 360 / circumference * titlePaint.measureText(title)
        val sweepAngle = (mSweepAngle - textAngle - 1 - 1) / 2
        if (isShowTitle) {
            // text left
            val leftStartAngle = mStartAngle
            canvas.drawArc(outlineOval!!, leftStartAngle, sweepAngle, false, outlinePaint)
            // text right
            val rightStartAngle = mStartAngle + mSweepAngle - sweepAngle
            canvas.drawArc(outlineOval!!, rightStartAngle, sweepAngle, false, outlinePaint)
        } else {
            canvas.drawArc(outlineOval!!, mStartAngle, mSweepAngle, false, outlinePaint)
        }
    }

    fun setMaxRate(maxRate: Int) {
        this.maxRate = maxRate
    }

    fun setStartAngle(mStartAngle: Float) {
        this.mStartAngle = mStartAngle
    }

    fun setSweepAngle(mSweepAngle: Float) {
        this.mSweepAngle = mSweepAngle
    }

    fun setCenterX(mCenterX: Int) {
        this.mCenterX = mCenterX
    }

    fun setCenterY(mCenterY: Int) {
        this.mCenterY = mCenterY
    }

    fun setIsSingle(isSingle: Boolean) {
        this.isSingle = isSingle
    }

    fun setRatedColor(color: Int) {
        ratedPaint.color = color
    }

    fun setUnRatedColor(color: Int) {
        unRatedPaint.color = color
    }

    fun setTitleColor(color: Int) {
        titlePaint.color = color
    }

    fun setOutlineColor(color: Int) {
        outlinePaint.color = color
    }

    fun setShadowColor(color: Int) {
        shadowPaint.color = color
    }

    fun isShowTitle(isShow: Boolean) {
        isShowTitle = isShow
    }

    fun setRatingBarColor(color: Int) {
        ratedPaint.color = color
        unRatedPaint.color = color
        titlePaint.color = color
        outlinePaint.color = color
        shadowPaint.color = color
    }

    /**
     * Rate class
     */
    inner class Rate(private val startAngle: Float, private val sweepAngle: Float) {
        fun drawArc(canvas: Canvas, oval: RectF?, paint: Paint?) {
            canvas.drawArc(oval!!, startAngle, sweepAngle, false, paint!!)
        }

    }

    companion object {
        private const val ITEM_OFFSET = 1
        private const val SHADOW_ALPHA = (255 * 0.2).toInt() // 20%
        private const val UN_RATING_ALPHA = (255 * 0.3).toInt() // 30%
        private const val RATING_ALPHA = (255 * 1.0).toInt() // 100%
        private const val OUTLINE_ALPHA = (255 * 0.4).toInt() // 40%
    }

    init {
        // set default color
        setRatingBarColor(Color.WHITE)
    }
}